File: backend/src/routes/work_entries.py
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query

from src.database import get_connection
from src.dependencies import get_authenticated_user
from src.schemas import CreateWorkEntryRequest, UpdateWorkEntryRequest

router = APIRouter(prefix="/api/work-entries", tags=["work-entries"])


def _fetch_entry_with_client(cur, entry_id: int) -> Optional[dict]:
    cur.execute(
        "SELECT we.id, we.client_id, we.hours, we.description, we.date, "
        "we.created_at, we.updated_at, c.name as client_name "
        "FROM work_entries we JOIN clients c ON we.client_id = c.id "
        "WHERE we.id = %s",
        (entry_id,),
    )
    return cur.fetchone()


@router.get("/")
def get_work_entries(
    clientId: Optional[int] = Query(default=None),
    user_email: str = Depends(get_authenticated_user),
):
    conn = get_connection()
    try:
        with conn.cursor() as cur:
            query = (
                "SELECT we.id, we.client_id, we.hours, we.description, we.date, "
                "we.created_at, we.updated_at, c.name as client_name "
                "FROM work_entries we JOIN clients c ON we.client_id = c.id "
                "WHERE we.user_email = %s"
            )
            params: list = [user_email]

            if clientId is not None:
                query += " AND we.client_id = %s"
                params.append(clientId)

            query += " ORDER BY we.date DESC, we.created_at DESC"
            cur.execute(query, params)
            return {"workEntries": cur.fetchall()}
    finally:
        conn.close()


@router.get("/{entry_id}")
def get_work_entry(entry_id: int, user_email: str = Depends(get_authenticated_user)):
    conn = get_connection()
    try:
        with conn.cursor() as cur:
            cur.execute(
                "SELECT we.id, we.client_id, we.hours, we.description, we.date, "
                "we.created_at, we.updated_at, c.name as client_name "
                "FROM work_entries we JOIN clients c ON we.client_id = c.id "
                "WHERE we.id = %s AND we.user_email = %s",
                (entry_id, user_email),
            )
            row = cur.fetchone()
            if not row:
                raise HTTPException(status_code=404, detail="Work entry not found")
            return {"workEntry": row}
    finally:
        conn.close()


@router.post("/", status_code=201)
def create_work_entry(body: CreateWorkEntryRequest, user_email: str = Depends(get_authenticated_user)):
    conn = get_connection()
    try:
        with conn.cursor() as cur:
            cur.execute(
                "SELECT id FROM clients WHERE id = %s AND user_email = %s",
                (body.clientId, user_email),
            )
            if not cur.fetchone():
                raise HTTPException(
                    status_code=400, detail="Client not found or does not belong to user"
                )

            cur.execute(
                "INSERT INTO work_entries (client_id, user_email, hours, description, date) "
                "VALUES (%s, %s, %s, %s, %s)",
                (body.clientId, user_email, body.hours, body.description, body.date),
            )
            new_id = cur.lastrowid
            row = _fetch_entry_with_client(cur, new_id)
            return {"message": "Work entry created successfully", "workEntry": row}
    finally:
        conn.close()


@router.put("/{entry_id}")
def update_work_entry(
    entry_id: int,
    body: UpdateWorkEntryRequest,
    user_email: str = Depends(get_authenticated_user),
):
    conn = get_connection()
    try:
        with conn.cursor() as cur:
            cur.execute(
                "SELECT id FROM work_entries WHERE id = %s AND user_email = %s",
                (entry_id, user_email),
            )
            if not cur.fetchone():
                raise HTTPException(status_code=404, detail="Work entry not found")

            data = body.model_dump(exclude_unset=True)
            if not data:
                raise HTTPException(status_code=400, detail="No fields to update")

            if "clientId" in data and data["clientId"] is not None:
                cur.execute(
                    "SELECT id FROM clients WHERE id = %s AND user_email = %s",
                    (data["clientId"], user_email),
                )
                if not cur.fetchone():
                    raise HTTPException(
                        status_code=400, detail="Client not found or does not belong to user"
                    )

            cur.execute(
                "SELECT client_id, hours, description, date FROM work_entries "
                "WHERE id = %s AND user_email = %s",
                (entry_id, user_email),
            )
            current = cur.fetchone()

            cur.execute(
                "UPDATE work_entries SET client_id = %s, hours = %s, description = %s, "
                "date = %s, updated_at = CURRENT_TIMESTAMP "
                "WHERE id = %s AND user_email = %s",
                (
                    data.get("clientId", current["client_id"]),
                    data.get("hours", current["hours"]),
                    data["description"] if "description" in data else current["description"],
                    data.get("date", current["date"]),
                    entry_id,
                    user_email,
                ),
            )
            row = _fetch_entry_with_client(cur, entry_id)
            return {"message": "Work entry updated successfully", "workEntry": row}
    finally:
        conn.close()


@router.delete("/{entry_id}")
def delete_work_entry(entry_id: int, user_email: str = Depends(get_authenticated_user)):
    conn = get_connection()
    try:
        with conn.cursor() as cur:
            cur.execute(
                "SELECT id FROM work_entries WHERE id = %s AND user_email = %s",
                (entry_id, user_email),
            )
            if not cur.fetchone():
                raise HTTPException(status_code=404, detail="Work entry not found")

            cur.execute(
                "DELETE FROM work_entries WHERE id = %s AND user_email = %s",
                (entry_id, user_email),
            )
            return {"message": "Work entry deleted successfully"}
    finally:
        conn.close()

File: backend/src/routes/workEntries.js
const express = require('express');
const { getDatabase } = require('../database/init');
const { authenticateUser } = require('../middleware/auth');
const { workEntrySchema, updateWorkEntrySchema } = require('../validation/schemas');
const {
  parseIdParam,
  validateIdParam,
  buildUpdateSet,
  handleDbError,
  WORK_ENTRY_SELECT,
} = require('./helpers');

/** Updatable work entry columns and their request-body keys. */
const UPDATE_FIELDS = [
  { column: 'client_id', key: 'clientId' },
  { column: 'hours', key: 'hours' },
  { column: 'description', key: 'description', nullable: true },
  { column: 'date', key: 'date' },
];

const router = express.Router();

// All routes require authentication
router.use(authenticateUser);

// Get all work entries for authenticated user (with optional client filter)
router.get('/', (req, res) => {
  const { clientId } = req.query;
  const db = getDatabase();
  
  let query = `${WORK_ENTRY_SELECT} WHERE we.user_email = ?`;
  
  const params = [req.userEmail];
  
  if (clientId) {
    const clientIdNum = parseIdParam(clientId);
    if (clientIdNum === null) {
      return res.status(400).json({ error: 'Invalid client ID' });
    }
    query += ' AND we.client_id = ?';
    params.push(clientIdNum);
  }
  
  query += ' ORDER BY we.date DESC, we.created_at DESC';
  
  db.all(query, params, (err, rows) => {
    if (err) {
      return handleDbError(res, err);
    }
    
    res.json({ workEntries: rows });
  });
});

// Get specific work entry
router.get('/:id', validateIdParam('work entry'), (req, res) => {
  const workEntryId = req.parsedId;
  const db = getDatabase();
  
  db.get(
    `${WORK_ENTRY_SELECT} WHERE we.id = ? AND we.user_email = ?`,
    [workEntryId, req.userEmail],
    (err, row) => {
      if (err) {
        return handleDbError(res, err);
      }
      
      if (!row) {
        return res.status(404).json({ error: 'Work entry not found' });
      }
      
      res.json({ workEntry: row });
    }
  );
});

// Create new work entry
router.post('/', (req, res, next) => {
  try {
    const { error, value } = workEntrySchema.validate(req.body);
    if (error) {
      return next(error);
    }

    const { clientId, hours, description, date } = value;
    const db = getDatabase();

    // Verify client exists and belongs to user
    db.get(
      'SELECT id FROM clients WHERE id = ? AND user_email = ?',
      [clientId, req.userEmail],
      (err, row) => {
        if (err) {
          return handleDbError(res, err);
        }

        if (!row) {
          return res.status(400).json({ error: 'Client not found or does not belong to user' });
        }

        // Create work entry
        db.run(
          'INSERT INTO work_entries (client_id, user_email, hours, description, date) VALUES (?, ?, ?, ?, ?)',
          [clientId, req.userEmail, hours, description || null, date],
          function(err) {
            if (err) {
              return handleDbError(res, err, 'Failed to create work entry');
            }

            // Return the created work entry with client name
            db.get(
              `${WORK_ENTRY_SELECT} WHERE we.id = ?`,
              [this.lastID],
              (err, row) => {
                if (err) {
                  return handleDbError(res, err, 'Work entry created but failed to retrieve');
                }

                res.status(201).json({
                  message: 'Work entry created successfully',
                  workEntry: row
                });
              }
            );
          }
        );
      }
    );
  } catch (error) {
    next(error);
  }
});

// Update work entry
router.put('/:id', validateIdParam('work entry'), (req, res, next) => {
  try {
    const workEntryId = req.parsedId;

    const { error, value } = updateWorkEntrySchema.validate(req.body);
    if (error) {
      return next(error);
    }

    const db = getDatabase();

    // Check if work entry exists and belongs to user
    db.get(
      'SELECT id FROM work_entries WHERE id = ? AND user_email = ?',
      [workEntryId, req.userEmail],
      (err, row) => {
        if (err) {
          return handleDbError(res, err);
        }

        if (!row) {
          return res.status(404).json({ error: 'Work entry not found' });
        }

        // If clientId is being updated, verify it belongs to user
        if (value.clientId) {
          db.get(
            'SELECT id FROM clients WHERE id = ? AND user_email = ?',
            [value.clientId, req.userEmail],
            (err, clientRow) => {
              if (err) {
                return handleDbError(res, err);
              }

              if (!clientRow) {
                return res.status(400).json({ error: 'Client not found or does not belong to user' });
              }

              performUpdate();
            }
          );
        } else {
          performUpdate();
        }

        function performUpdate() {
          const { setClause, values } = buildUpdateSet(UPDATE_FIELDS, value);
          values.push(workEntryId, req.userEmail);

          const query = `UPDATE work_entries SET ${setClause} WHERE id = ? AND user_email = ?`;

          db.run(query, values, function(err) {
            if (err) {
              return handleDbError(res, err, 'Failed to update work entry');
            }

            // Return updated work entry with client name
            db.get(
              `${WORK_ENTRY_SELECT} WHERE we.id = ?`,
              [workEntryId],
              (err, row) => {
                if (err) {
                  return handleDbError(res, err, 'Work entry updated but failed to retrieve');
                }

                res.json({
                  message: 'Work entry updated successfully',
                  workEntry: row
                });
              }
            );
          });
        }
      }
    );
  } catch (error) {
    next(error);
  }
});

// Delete work entry
router.delete('/:id', validateIdParam('work entry'), (req, res) => {
  const workEntryId = req.parsedId;
  const db = getDatabase();
  
  // Check if work entry exists and belongs to user
  db.get(
    'SELECT id FROM work_entries WHERE id = ? AND user_email = ?',
    [workEntryId, req.userEmail],
    (err, row) => {
      if (err) {
        return handleDbError(res, err);
      }
      
      if (!row) {
        return res.status(404).json({ error: 'Work entry not found' });
      }
      
      // Delete work entry
      db.run(
        'DELETE FROM work_entries WHERE id = ? AND user_email = ?',
        [workEntryId, req.userEmail],
        function(err) {
          if (err) {
            return handleDbError(res, err, 'Failed to delete work entry');
          }
          
          res.json({ message: 'Work entry deleted successfully' });
        }
      );
    }
  );
});

module.exports = router;
